lighthouse-viewer/app/src/main.js

```diff
@@ -0,0 +1,504 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+/* global window, document */
+
+const ReportGenerator = require('../../../lighthouse-core/report/report-generator');
+// TODO: We only need getFilenamePrefix from asset-saver. Tree shake!
+const getFilenamePrefix = require('../../../lighthouse-core/lib/asset-saver.js').getFilenamePrefix;
+
+const firebase = require('firebase/app');
+require('firebase/auth');
+const idb = require('idb-keyval');
+
+// Polyfills
+// TODO: dynamically load this based on feature detection.
+require('whatwg-fetch');
+const URLSearchParams = require('url-search-params');
+
+const LH_CURRENT_VERSION = require('../../../package.json').version;
+const APP_URL = `${location.origin}${location.pathname}`;
+
+class Logger {
+  constructor(selector) {
+    this.el = document.querySelector(selector);
+  }
+
+  /**
+   * Shows a butter bar.
+   * @param {!string} msg The message to show.
+   * @param {boolean=} autoHide True to hide the message after a duration.
+   *     Default is true.
+   */
+  log(msg, autoHide = true) {
+    clearTimeout(this._id);
+
+    this.el.textContent = msg;
+    this.el.classList.add('show');
+    if (autoHide) {
+      this._id = setTimeout(() => {
+        this.el.classList.remove('show');
+      }, 7000);
+    }
+  }
+
+  warn(msg) {
+    console.warn(msg);
+  }
+
+  error(msg) {
+    console.error(msg);
+  }
+
+  /**
+   * Explicitly hides the butter bar.
+   */
+  hide() {
+    clearTimeout(this._id);
+    this.el.classList.remove('show');
+  }
+}
+
+const logger = new Logger('#log');
+
+class FirebaseAuth {
+  constructor() {
+    this.accessToken = null;
+    this.user = null;
+
+    this.provider = new firebase.auth.GithubAuthProvider();
+    this.provider.addScope('gist');
+
+    firebase.initializeApp({
+      apiKey: 'AIzaSyApMz8FHTyJNqqUtA51tik5Mro8j-2qMcM',
+      authDomain: 'lighthouse-viewer.firebaseapp.com',
+      databaseURL: 'https://lighthouse-viewer.firebaseio.com',
+      storageBucket: 'lighthouse-viewer.appspot.com',
+      messagingSenderId: '962507201498'
+    });
+
+    // Wrap auth state callback in a promise so  other parts of the app that
+    // require login can hook into the changes.
+    this.ready = new Promise((resolve, reject) => {
+      firebase.auth().onAuthStateChanged(user => {
+        idb.get('accessToken').then(token => {
+          if (user && token) {
+            this.accessToken = token;
+            this.user = user;
+          }
+          resolve(user);
+        });
+      });
+    });
+  }
+
+  /**
+   * Signs in the user to Github using the Firebase API.
+   * @return {!Promise<object>} The logged in user.
+   */
+  signIn() {
+    return firebase.auth().signInWithPopup(this.provider).then(result => {
+      this.accessToken = result.credential.accessToken;
+      // A limitation of FB auth is that it doesn't return an oauth token
```

Comment:
is "FB" the normal abbreviation here? Thought this was somehow facebook auth for a second :)